openeo/extra/job_management/job_splitting.py

```diff
@@ -0,0 +1,136 @@
+        return cls(*polygon.bounds, projection if projection is not None else cls.crs)
+
+    def as_dict(self) -> Dict:
+        return self._asdict()
+
+    def as_polygon(self) -> shapely.geometry.Polygon:
+        """Get bounding box as a shapely Polygon"""
+        return shapely.geometry.box(minx=self.west, miny=self.south, maxx=self.east, maxy=self.north)
+
+
+class TileGridInterface(metaclass=abc.ABCMeta):
+    """Interface for tile grid classes"""
+
+    @abc.abstractmethod
+    def get_tiles(self, geometry: Union[Dict, shapely.geometry.Polygon]) -> List[Union[Dict, shapely.geometry.Polygon]]:
+        """Calculate tiles to cover given bounding box"""
+        ...
+
+
+class SizeBasedTileGrid(TileGridInterface):
+    """
+    Specification of a tile grid, parsed from a size and a projection.
+    """
+
+    def __init__(self, epsg: str, size: float):
+        self.epsg = epsg
+        self.size = size
+
+    @classmethod
+    def from_size_projection(cls, size: float, projection: str) -> "SizeBasedTileGrid":
+        """Create a tile grid from size and projection"""
+        return cls(projection.lower(), size)
+
+    def get_tiles(self, geometry: Union[Dict, shapely.geometry.Polygon]) -> List[Union[Dict, shapely.geometry.Polygon]]:
+        if isinstance(geometry, dict):
+            bbox = BoundingBox.from_dict(geometry)
+            bbox_crs = bbox.crs
+        elif isinstance(geometry, shapely.geometry.Polygon):
+            bbox = BoundingBox.from_polygon(geometry, projection=self.epsg)
+            bbox_crs = self.epsg
+        else:
+            raise JobSplittingFailure("geometry must be a dict or a shapely.geometry.Polygon")
+
+        if self.epsg == "epsg:4326":
+            tile_size = self.size
+            x_offset = 0
+        else:
+            tile_size = self.size * 1000
+            x_offset = 500_000
+
+        to_cover = BoundingBox.from_dict(reproject_bounding_box(bbox.as_dict(), from_crs=bbox_crs, to_crs=self.epsg))
+        xmin = int(math.floor((to_cover.west - x_offset) / tile_size))
+        xmax = int(math.ceil((to_cover.east - x_offset) / tile_size)) - 1
+        ymin = int(math.floor(to_cover.south / tile_size))
+        ymax = int(math.ceil(to_cover.north / tile_size)) - 1
+
+        tiles = []
+        for x in range(xmin, xmax + 1):
+            for y in range(ymin, ymax + 1):
+                tile = BoundingBox(
+                    west=max(x * tile_size + x_offset, to_cover.west),
+                    south=max(y * tile_size, to_cover.south),
+                    east=min((x + 1) * tile_size + x_offset, to_cover.east),
+                    north=min((y + 1) * tile_size, to_cover.north),
+                    crs=self.epsg,
+                )
+
+                if isinstance(geometry, dict):
+                    tiles.append(reproject_bounding_box(tile.as_dict(), from_crs=self.epsg, to_crs=bbox_crs))
+                else:
+                    tiles.append(tile.as_polygon())
```

Comment:
This looks like risky behavior if I understand correctly: in one case you reproject and in another you don't. I'm not sure users will anticipate that when naively using this function.
I understand you can document that, but these kind of important details are easily glossed over.

Reply:
at first sight I think this method tries to do too much (bboxes with crs, polygons without crs ..., reprojections, ...) 

I'd extract the core functionality (double for-loop) as a helper and make its logic "CRS-free". And reuse that in more specialized versions (for CRS-aware bboxes, for naive polygons, ...)

Reply:
yes I would make sure to have a core function that does just tile splitting in a naive way (in the sense that it does not considers any CRS aspects). This function would then not have to depend on pyproj. 
Then, reuse this function for particular uses cases that possibly require reprojects, which is fine as long as it does not leak into the core splitting function.
Not sure if it's clear what I mean :)